Fix for variable dependency bug when initializing stateful properties.

aeslaughter opened the pull request:
@WilkAndy The first commit should allow you to run your test w/o the "yaqi_hacky" setting.

@idaholab/moose-team The second commit I am attempting to create a test for this fix, but I am unable to reproduce the problem in MOOSE itself. Does anyone have any ideas? Should I even worry about it?

aeslaughter committed:
Add missing variable dependency call for init of stateful properties

(closes #6774)

aeslaughter committed:
Creates a test that exercises the dependency problem

(refs #6774)

andrsd commented on framework/src/base/MooseVariable.C:
```diff
@@ -175,12 +175,6 @@ MooseVariable::prepare()
-  if (_dof_indices.size() > 0)
-    _is_defined = true;
-  else
-    _is_defined = false;
```
👍

andrsd commented:
I want to take a look at the test today...

andrsd commented on framework/src/base/ComputeMaterialsObjectThread.C:
```diff
@@ -76,6 +77,10 @@ ComputeMaterialsObjectThread::subdomainChanged()
+
+  std::set<MooseVariable *> needed_moose_vars;
+  _materials.updateVariableDependency(needed_moose_vars, _tid);
+  //_fe_problem.setActiveElementalMooseVariables(needed_moose_vars, _tid);
```
We will want to remove this line in the final version...

andrsd replied:
I am wrong here. We need to uncomment the line ;-)

andrsd commented:
Figured out the test and uncommented the `_fe_problem` line. It all sits in [andrsd/7116](https://github.com/andrsd/moose/tree/7116). You can fetch, merge and interactive autosquash my branch...

aeslaughter commented:
Thanks @andrsd for figuring out the test for this problem.

@WilkAndy This should fix your problem so you no longer need the "yaqi_hacky" lines.

aeslaughter committed:
Make warehouse variable depenedency handler more generic

aeslaughter committed:
Update variable dependencies prior to running general userobjects

aeslaughter committed:
Added variable dependency interface to ElementVariableValue

aeslaughter committed:
Added missing dependency interface to FindValueOnLine

(refs #6774)

WilkAndy commented:
Great, i'll submit my yaqi_hacky removal - #7126 - when this gets merged.  Thanks guys,

a

aeslaughter commented:
@WilkAndy It might be a while before I get this merged. This issue uncovered a widespread deficiency in how we handle variable dependencies. I have it working for MOOSE but all sorts of modules test fail.

WilkAndy commented:
That's fantastic really.  i can still work with my yaqi_hacky and we get to make the code more robust

WilkAndy commented:
Hi @aeslaughter .  When this is nearing completion, could you activate the heavy tests?  i have quite a few that i don't want failing.

aeslaughter committed:
Update FeatureFloodCount to set variable dependencies

(refs #6774)

aeslaughter committed:
Make it an error to inherit from MooseVariableDependencyInterface w/o calling the add method

(refs #6774)

aeslaughter committed:
Updated variable dependency calls so that no error is produced

aeslaughter committed:
Update contraints to use proper dependency calls

aeslaughter committed:
Updates to migrate to complete integration with variable dependency interface

aeslaughter committed:
MooseVariable dependency complete for moose; Removes the need for preparing all variables by default

aeslaughter committed:
Updates the clear statements to match set statements; should allow for existing applications to work w/o modification

This also includes a mooseWarning if you end up in the general prepare all element methods, telling people to set the dependencies.
(refs #6774)

aeslaughter committed:
Why are you always preparing...just go

aeslaughter commented:
@bwspenc There is a single modules test failing in this PR, base on my examination it is possible that the gold may be wrong, but I don't know anything about this test, would you be willing to take a look?

In solid mechancs:
```
line_material_symm_tensor_sampler.test...................................................... FAILED (CSVDIFF) 
```

dschwen commented:
@hchen139 is currently porting this to tensor mechanics and we discussed whether to remove this object and test from solid mechanics entirely.

bwspenc commented:
@aeslaughter Sure, I can take a look, although I'll have to get up to speed on this yaqi_hacky stuff.  @dschwen is right -- we want to remove this capability.  It was developed before we had vector postprocessors, and we should have removed it long ago. The easiest way to fix this is to delete the test, and deprecate that code if we're OK with having deprecated code be untested.

bwspenc commented:
I take that back -- I think this is actually a test of the new stuff.  It's something else that we wanted to deprecate.  I'll take a look.

dschwen commented:
Yeah, the thing that fails is a VPP. It is likely that the TM version will
fail as well (same base class).

On Fri, Jun 3, 2016, 10:00 AM Ben Spencer <notifications@github.com> wrote:

> I take that back -- I think this is actually a test of the new stuff. It's
> something else that we wanted to deprecate. I'll take a look.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7116#issuecomment-223619453>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAMWPiHL9ldTltjN-IpJOooENhJzGmhsks5qIE-EgaJpZM4Iq0FU>
> .
>

hchen139 commented:
@dschwen @bwspenc the failed test in this PR will be fixed in PR #7092, which use the stress_xx auxvariable not previously used in line_material_symm_tensor_test to test the LineMaterialRankTwoSampler VPP

andrsd commented:
@aeslaughter I rebased this branch on top of current `devel`, it is in [andrsd/pr-7116](https://github.com/andrsd/moose/tree/pr-7116). 2 tests are still failing in modules - but only exodiffs. I do not know if it related or not to these changes...

andrsd commented:
One of the failing tests is still `line_material_symm_tensor_sampler`... The other one is `generalized_plane_strain_tm_contact.generalized_plane_strain_tm_contact_field`

aeslaughter commented:
I can take a look again, I forgot all about this branch and it will take me some time to recall what I was doing.

aeslaughter committed:
Minor changes from rebase on devel

aeslaughter commented:
@andrsd I was able to re-base this branch and am getting the same failures that you mention, I will see if I can figure them out.

```
solid_mechanics:line_material_symm_tensor_sampler.test...................................... FAILED (CSVDIFF)
tensor_mechanics:line_material_rank_two_sampler.rank_two_sampler............................ FAILED (CSVDIFF)
tensor_mechanics:line_material_rank_two_sampler.rank_two_scalar_sampler..................... FAILED (CSVDIFF)
```

aeslaughter commented:
3bde916 is the problem with these tests.

```
3bde916b3826269f679c45b8c6f1a8ae64def091 is the first bad commit
commit 3bde916b3826269f679c45b8c6f1a8ae64def091
Author: Andrew E Slaughter <andrew.slaughter@inl.gov>
Date:   Wed Jun 1 15:52:58 2016 -0600

    Update variable dependencies prior to running general userobjects

:040000 040000 1a826b909bbdc8f1d93aae4b4784b8ccddb5b620 8858879c9109f0adfd98761a5de7ec2203b9cf74 M	framework
```

aeslaughter committed:
Improve MooseVariable dependency to automatically include dependencies from Material properties consumed by an object

(refs #6774)

aeslaughter committed:
Update MooseVariableDependencyInterface to accept MooseObject pointer

(refs #6774)

aeslaughter commented:
I have a patch for Bighorn for these changes.

@YaqiWang @snschune I need some help with YAK/RattleSnake

snschune commented:
Lol `Output too large. You will need to download the results to see this.` How do I download this?

permcody commented:
Click on the big blue arrow next to the "Results" header.

On Tue, Nov 15, 2016 at 3:22 PM Sebastian Schunert <notifications@github.com>
wrote:

> Lol Output too large. You will need to download the results to see this.
> How do I download this?
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7116#issuecomment-260789301>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIOVILM9BbUNMRtPCnRvACgNX-Ctlks5q-jCjgaJpZM4Iq0FU>
> .
>

aeslaughter committed:
Changes to due to name change with FEProblem and NonlinearSystem

(refs #6774)

aeslaughter commented:
@permcody I am not sure if this is related to what you had @lindsayad, but I did this work to improve variable dependency a few months ago. Of course a few YAK tests were failing and never resolved so this had been sitting around. I just did a rebase, lets see what happens.

lindsayad commented:
@aeslaughter Failing tests related to `LineMaterial`...I had the same issue. I don't know if this is the same here, but I had to add `getMatPropDependencies` and `setActiveMaterialProperties` calls before `reinitMaterials` in `LineMaterialSamplerBase::execute` so that my material properties were actually computed when they needed to be. Maybe a similar thing needs to be done for variable dependencies before `reinitElem` is called? This could be an issue for anything that inherits from `GeneralVectorPostprocessor` and `GeneralUserObject` that has variable and/or material dependencies.

permcody commented:
Yes Alex, it's rare for General objects to access Material properties. Usually this happens because someone needs control over the element or nodal loops so they still want that information. In these few cases, it's reasonable to require that these objects inherit from the MaterialPropertyInterface or otherwise take care of reporting their dependencies to the system.

aeslaughter committed:
Changes due to rebasing with current MOOSE

(refs #6774)

aeslaughter commented:
@permcody This interface is currently not using a MooseObject, the deprecated message if new.

https://github.com/idaholab/moose/blob/513f0cb4da8102d80ad00a5474f9f88955ef2152/framework/src/userobject/SideUserObject.C#L34

permcody commented:
I'm not sure why this PR is open, @aeslaughter? I think this fix has been merged in. At any rate, it's super stale and should probably be closed. @dschwen?

aeslaughter commented:
This has never been merged, I was never able to get it to pass YAK. This should be re-visisted sometime.

snschune commented:
The problem might have been solved with the changes for material dependency resolution. I vaguely remember the same set of tests failing: objects using materials that are not executed within threaded loops.

dschwen commented:
Stale. Feel free to reopen if this is still needed.
